internal/mode/static/nginx/config/servers_template.go

```diff
@@ -46,20 +47,102 @@ server {
         js_content httpmatches.redirect;
         {{- end }}
 
+        {{ $proxyOrGRPC := "proxy" }}{{ if $l.IsGRPC }}{{ $proxyOrGRPC = "grpc" }}{{ end }}
+
+        {{- if $l.IsGRPC }}
+        error_page 400 = @grpc_internal;
+        error_page 401 = @grpc_unauthenticated;
+        error_page 403 = @grpc_permission_denied;
+        error_page 404 = @grpc_unimplemented;
+        error_page 429 = @grpc_unavailable;
+        error_page 502 = @grpc_unavailable;
+        error_page 503 = @grpc_unavailable;
+        error_page 504 = @grpc_unavailable;
+        error_page 405 = @grpc_internal;
+        error_page 408 = @grpc_deadline_exceeded;
+        error_page 413 = @grpc_resource_exhausted;
+        error_page 414 = @grpc_resource_exhausted;
+        error_page 415 = @grpc_internal;
+        error_page 426 = @grpc_internal;
+        error_page 495 = @grpc_unauthenticated;
+        error_page 496 = @grpc_unauthenticated;
+        error_page 497 = @grpc_internal;
+        error_page 500 = @grpc_internal;
+        error_page 501 = @grpc_internal;
+        {{- end }}
+
         {{- if $l.ProxyPass -}}
             {{ range $h := $l.ProxySetHeaders }}
-        proxy_set_header {{ $h.Name }} "{{ $h.Value }}";
+        {{ $proxyOrGRPC }}_set_header {{ $h.Name }} "{{ $h.Value }}";
             {{- end }}
+        {{ $proxyOrGRPC }}_pass {{ $l.ProxyPass }};
         proxy_http_version 1.1;
-        proxy_pass {{ $l.ProxyPass }};
             {{- if $l.ProxySSLVerify }}
-        proxy_ssl_verify on;
-        proxy_ssl_name {{ $l.ProxySSLVerify.Name }};
-        proxy_ssl_trusted_certificate {{ $l.ProxySSLVerify.TrustedCertificate }};
+        {{ $proxyOrGRPC }}_ssl_verify on;
+        {{ $proxyOrGRPC }}_ssl_name {{ $l.ProxySSLVerify.Name }};
+        {{ $proxyOrGRPC }}_ssl_trusted_certificate {{ $l.ProxySSLVerify.TrustedCertificate }};
             {{- end }}
         {{- end }}
     }
         {{ end }}
+
+        {{- if $s.HTTP2 }}
+    location @grpc_deadline_exceeded {
```

Comment:
To avoid repetitions, we can put error_page directives into a separate file as well as locations.
And include it in the config:
```
location {
  . . .
  include grpc-error-pages.conf;
  . . .
}
. . .
include grpc-error-locations.conf;
. . .
```